File: toggly-ruby/toggly/lib/toggly/telemetry/grpc_clients.rb
# frozen_string_literal: true

require "uri"

module Toggly
  module Telemetry
    # Optional gRPC transport helpers for usage and metrics telemetry.
    module GrpcClients
      DEFAULT_METRICS_BASE_URL = "https://app.toggly.io/"
      DEFAULT_TELEMETRY_FLUSH_SECONDS = 60.0

      # HTTP/2 metadata is case-insensitive; .NET/Go/Node send ``UA``.
      # The Ruby grpc gem lowercases keys, so we send ``ua`` with the same
      # user-agent semantics.
      GRPC_USER_AGENT_METADATA_KEY = "ua"

      module_function

      # FNV-1a 32-bit as signed int32 (UTF-8 bytes; matches Go/Node/Python).
      #
      # @param identity [String]
      # @return [Integer] signed int32
      def hash_identity(identity)
        h = 2_166_136_261
        identity.to_s.encode("UTF-8").bytes.each do |byte|
          h ^= byte
          h = (h * 16_777_619) & 0xFFFFFFFF
        end
        h > 0x7FFFFFFF ? h - 0x100000000 : h
      end

      # @param time [Time, nil]
      # @return [Hash{Symbol => Integer}] protobuf Timestamp shape
      def to_protobuf_timestamp(time = nil)
        t = time || Time.now.utc
        t = t.utc
        ms = (t.to_f * 1000).to_i
        {
          seconds: ms / 1000,
          nanos: (ms % 1000) * 1_000_000
        }
      end

      # Build a protobuf Timestamp from a batcher hash. Nested message fields are
      # nil until assigned — callers must assign the returned object, not mutate
      # ``msg.time.seconds``.
      #
      # @param timestamp_hash [Hash, nil]
      # @return [Google::Protobuf::Timestamp, nil]
      def build_timestamp(timestamp_hash)
        return nil unless timestamp_hash.is_a?(Hash)

        require "google/protobuf/timestamp_pb"
        Google::Protobuf::Timestamp.new(
          seconds: timestamp_hash[:seconds].to_i,
          nanos: timestamp_hash[:nanos].to_i
        )
      end

      # @param base_url [String]
      # @return [String] host:port for a gRPC channel
      def grpc_target(base_url)
        raw = base_url.to_s.strip
        raw = "https://#{raw}" unless raw.include?("://")
        uri = begin
          URI.parse(raw)
        rescue URI::InvalidURIError
          nil
        end
        host = if uri
                 uri.host || uri.path
               else
                 raw.sub(%r{\Ahttps?://}i, "").delete_suffix("/")
               end
        host = host.to_s.delete_suffix("/")
        host = "app.toggly.io:443" if host.empty?
        host = "#{host}:443" unless host.include?(":")
        host
      end

      # @param override [String, nil]
      # @return [String]
      def resolve_user_agent(override = nil)
        override || "toggly-ruby/#{Toggly::VERSION}"
      end

      # @return [Boolean]
      def grpc_available?
        return @grpc_available unless @grpc_available.nil?

        require "grpc"
        require_relative "pb/usage_services_pb"
        require_relative "pb/metrics_services_pb"
        @grpc_available = true
      rescue LoadError
        @grpc_available = false
      end

      # Reset cached availability (tests only).
      def reset_grpc_available!
        @grpc_available = nil
      end

      # Convert a batcher dict into Usage.FeatureStat.
      #
      # @param payload [Hash]
      # @return [Toggly::Telemetry::Pb::Usage::FeatureStat]
      def feature_stat_from_payload(payload)
        NativeUsageClient.feature_stat_from_payload(payload)
      end

      # Convert a batcher dict into Metrics.MetricStat.
      #
      # @param payload [Hash]
      # @return [Toggly::Telemetry::Pb::Metrics::MetricStat]
      def metric_stat_from_payload(payload)
        NativeMetricsClient.metric_stat_from_payload(payload)
      end

      # @param metrics_base_url [String]
      # @param user_agent [String, nil]
      # @param timeout [Numeric]
      # @return [Clients, nil]
      def create(metrics_base_url, user_agent: nil, timeout: 10.0)
        return nil unless grpc_available?

        require "grpc"

        target = grpc_target(metrics_base_url)
        channel = GRPC::Core::Channel.new(target, {}, GRPC::Core::ChannelCredentials.new)
        shared = SharedChannel.new(channel)
        ua = resolve_user_agent(user_agent)
        default_meta = { GRPC_USER_AGENT_METADATA_KEY => ua }

        usage_stub = Pb::Usage::Usage::Stub.new(target, GRPC::Core::ChannelCredentials.new,
                                                channel_override: channel)
        metrics_stub = Pb::Metrics::Metrics::Stub.new(target, GRPC::Core::ChannelCredentials.new,
                                                      channel_override: channel)

        Clients.new(
          usage: NativeUsageClient.new(usage_stub, shared, default_meta, timeout),
          metrics: NativeMetricsClient.new(metrics_stub, shared, default_meta, timeout)
        )
      end

      # Paired usage + metrics clients.
      Clients = Struct.new(:usage, :metrics, keyword_init: true)

      # Shared channel closer (close once).
      class SharedChannel
        def initialize(channel)
          @channel = channel
          @closed = false
          @mutex = Mutex.new
        end

        def close
          @mutex.synchronize do
            return if @closed

            @closed = true
            @channel.close
          end
        end
      end

      # Converts batcher hashes into Usage.FeatureStat and sends via stub.
      class NativeUsageClient
        def initialize(stub, shared, default_metadata, timeout)
          @stub = stub
          @shared = shared
          @default_metadata = default_metadata.transform_keys(&:to_s)
          @timeout = timeout
        end

        def send_stats(request, metadata: nil)
          meta = @default_metadata.merge((metadata || {}).transform_keys(&:to_s))
          msg = self.class.feature_stat_from_payload(request)
          @stub.send_stats(msg, metadata: meta.to_a, deadline: Time.now + @timeout)
        end

        def close
          @shared.close
        end

        # @param payload [Hash]
        # @return [Toggly::Telemetry::Pb::Usage::FeatureStat]
        def self.feature_stat_from_payload(payload)
          require_relative "pb/usage_pb"

          msg = Pb::Usage::FeatureStat.new(
            appKey: payload[:appKey].to_s,
            environment: payload[:environment].to_s,
            totalUniqueUsers: payload[:totalUniqueUsers].to_i,
            uniqueUserHashes: Array(payload[:uniqueUserHashes]).map(&:to_i)
          )
          apply_feature_stat_metadata(msg, payload)
          Array(payload[:stats]).each do |stat|
            next unless stat.is_a?(Hash)

            msg.stats << stat_message_from_hash(stat)
          end
          msg
        end

        def self.apply_feature_stat_metadata(msg, payload)
          timestamp = GrpcClients.build_timestamp(payload[:time])
          msg.time = timestamp if timestamp
          msg.instanceName = payload[:instanceName].to_s if payload[:instanceName]
          msg.appVersion = payload[:appVersion].to_s if payload[:appVersion]
          process_start = GrpcClients.build_timestamp(payload[:processStartTime])
          msg.processStartTime = process_start if process_start
          msg.definitionCacheHits = payload[:definitionCacheHits].to_i if payload[:definitionCacheHits]
          msg.definitionCacheMisses = payload[:definitionCacheMisses].to_i if payload[:definitionCacheMisses]
        end
        private_class_method :apply_feature_stat_metadata

        def self.stat_message_from_hash(stat)
          sm = Pb::Usage::StatMessage.new(
            feature: stat[:feature].to_s,
            uniqueContextIdentifierEnabledCount: stat[:uniqueContextIdentifierEnabledCount].to_i,
            uniqueContextIdentifierDisabledCount: stat[:uniqueContextIdentifierDisabledCount].to_i,
            uniqueUsersUsedCount: stat[:uniqueUsersUsedCount].to_i,
            uniqueUserHashes: Array(stat[:uniqueUserHashes]).map(&:to_i),
            uniqueViewedUserHashes: Array(stat[:uniqueViewedUserHashes]).map(&:to_i)
          )
          (stat[:variantStats] || {}).each do |name, vs|
            next unless vs.is_a?(Hash)

            sm.variantStats[name.to_s] = Pb::Usage::VariantStats.new(
              checkCount: vs[:checkCount].to_i,
              requestCount: vs[:requestCount].to_i,
              usedCount: vs[:usedCount].to_i,
              viewedCount: vs[:viewedCount].to_i
            )
          end
          sm
        end
        private_class_method :stat_message_from_hash
      end

      # Converts batcher hashes into Metrics.MetricStat and sends via stub.
      class NativeMetricsClient
        def initialize(stub, shared, default_metadata, timeout)
          @stub = stub
          @shared = shared
          @default_metadata = default_metadata.transform_keys(&:to_s)
          @timeout = timeout
        end

        def send_metrics(request, metadata: nil)
          meta = @default_metadata.merge((metadata || {}).transform_keys(&:to_s))
          msg = self.class.metric_stat_from_payload(request)
          @stub.send_metrics(msg, metadata: meta.to_a, deadline: Time.now + @timeout)
        end

        def close
          @shared.close
        end

        # @param payload [Hash]
        # @return [Toggly::Telemetry::Pb::Metrics::MetricStat]
        def self.metric_stat_from_payload(payload)
          require_relative "pb/metrics_pb"

          msg = Pb::Metrics::MetricStat.new(
            appKey: payload[:appKey].to_s,
            environment: payload[:environment].to_s
          )
          apply_metric_stat_metadata(msg, payload)
          append_metric_stats(msg, payload[:stats])
          append_metric_counters(msg, payload[:counters])
          append_metric_observations(msg, payload[:observations])
          msg
        end

        def self.apply_metric_stat_metadata(msg, payload)
          timestamp = GrpcClients.build_timestamp(payload[:time])
          msg.time = timestamp if timestamp
          msg.instanceName = payload[:instanceName].to_s if payload[:instanceName]
        end
        private_class_method :apply_metric_stat_metadata

        def self.append_metric_stats(msg, items)
          Array(items).each do |item|
            next unless item.is_a?(Hash)

            sm = Pb::Metrics::MetricStatMessage.new(metric: item[:metric].to_s)
            sm.feature = item[:feature].to_s if item[:feature]
            fill_variant_values(sm, item[:variantValues])
            msg.stats << sm
          end
        end
        private_class_method :append_metric_stats

        def self.append_metric_counters(msg, items)
          Array(items).each do |item|
            next unless item.is_a?(Hash)

            cm = Pb::Metrics::MetricCounterMessage.new(metric: item[:metric].to_s)
            cm.feature = item[:feature].to_s if item[:feature]
            fill_variant_values(cm, item[:variantValues])
            msg.counters << cm
          end
        end
        private_class_method :append_metric_counters

        def self.append_metric_observations(msg, items)
          Array(items).each do |item|
            next unless item.is_a?(Hash)

            om = Pb::Metrics::MetricObservationMessage.new(metric: item[:metric].to_s)
            obs_ts = GrpcClients.build_timestamp(item[:time])
            om.time = obs_ts if obs_ts
            om.feature = item[:feature].to_s if item[:feature]
            fill_variant_values(om, item[:variantValues])
            msg.observations << om
          end
        end
        private_class_method :append_metric_observations

        def self.fill_variant_values(target, variant_values)
          return unless variant_values.is_a?(Hash)

          variant_values.each do |name, value|
            target.variantValues[name.to_s] = value.to_f
          end
        end
        private_class_method :fill_variant_values
      end
    end
  end
end

File: Toggly.FeatureManagement.Python/toggly/src/toggly/telemetry/grpc_clients.py
"""Optional gRPC transport helpers for usage and metrics telemetry."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable, Mapping, Optional, Protocol
from urllib.parse import urlparse

from toggly.version import __version__

logger = logging.getLogger("toggly.telemetry")

DEFAULT_METRICS_BASE_URL = "https://app.toggly.io/"
DEFAULT_TELEMETRY_FLUSH_SECONDS = 60.0

# HTTP/2 metadata is case-insensitive; .NET/Go/Node send ``UA``.
# grpcio requires lowercase ASCII keys, so the Python client uses ``ua``
# with the same user-agent semantics.
GRPC_USER_AGENT_METADATA_KEY = "ua"


def hash_identity(identity: str) -> int:
    """FNV-1a 32-bit as signed int32 (UTF-8 bytes; matches Go/Node)."""
    h = 2166136261
    for byte in identity.encode("utf-8"):
        h ^= byte
        h = (h * 16777619) & 0xFFFFFFFF
    return h - 0x100000000 if h > 0x7FFFFFFF else h


def to_protobuf_timestamp(dt: datetime | None = None) -> dict[str, int]:
    """Return ``{seconds, nanos}`` for a protobuf Timestamp."""
    if dt is None:
        dt = datetime.now(timezone.utc)
    elif dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    else:
        dt = dt.astimezone(timezone.utc)
    ms = int(dt.timestamp() * 1000)
    seconds = ms // 1000
    nanos = (ms % 1000) * 1_000_000
    return {"seconds": seconds, "nanos": nanos}


def grpc_target(base_url: str) -> str:
    """Derive ``host:port`` for a gRPC channel from a metrics base URL."""
    raw = base_url.strip()
    if "://" not in raw:
        raw = f"https://{raw}"
    try:
        parsed = urlparse(raw)
        host = parsed.netloc or parsed.path
    except Exception:
        host = raw.replace("https://", "").replace("http://", "").rstrip("/")
    host = host.rstrip("/")
    if not host:
        host = "app.toggly.io:443"
    if ":" not in host:
        host = f"{host}:443"
    return host


def resolve_user_agent(override: str | None = None) -> str:
    """Return gRPC user-agent metadata value (``ua`` key; matches .NET/Go UA)."""
    return override or f"toggly-python/{__version__}"


def is_grpc_available() -> bool:
    """Return True when optional grpcio + generated stubs can be imported."""
    try:
        import grpc  # noqa: F401
        from google.protobuf import timestamp_pb2  # noqa: F401

        from toggly.telemetry.pb import (  # noqa: F401  # noqa: F401
            metrics_pb2,
            metrics_pb2_grpc,
            usage_pb2,
            usage_pb2_grpc,
        )

        return True
    except Exception:
        return False


class UsageGrpcClient(Protocol):
    """Minimal usage transport used by :class:`TelemetryRuntime`."""

    def send_stats(
        self,
        request: Mapping[str, Any],
        metadata: Mapping[str, str] | None = None,
    ) -> Any:
        """Send a FeatureStat payload."""

    def close(self) -> None:
        """Close the underlying channel/stub."""


class MetricsGrpcClient(Protocol):
    """Minimal metrics transport used by :class:`TelemetryRuntime`."""

    def send_metrics(
        self,
        request: Mapping[str, Any],
        metadata: Mapping[str, str] | None = None,
    ) -> Any:
        """Send a MetricStat payload."""

    def close(self) -> None:
        """Close the underlying channel/stub."""


@dataclass
class GrpcClients:
    """Paired usage + metrics gRPC clients."""

    usage: UsageGrpcClient
    metrics: MetricsGrpcClient


def _timestamp_message(ts: Mapping[str, int]) -> Any:
    from google.protobuf.timestamp_pb2 import Timestamp

    msg = Timestamp()
    msg.seconds = int(ts.get("seconds", 0))
    msg.nanos = int(ts.get("nanos", 0))
    return msg


def feature_stat_from_payload(payload: Mapping[str, Any]) -> Any:
    """Convert a batcher dict payload into a Usage.FeatureStat message."""
    from toggly.telemetry.pb import usage_pb2

    # Generated pb2 modules are dynamically typed for mypy.
    msg = usage_pb2.FeatureStat()  # type: ignore[attr-defined]
    msg.appKey = str(payload.get("appKey", ""))
    msg.environment = str(payload.get("environment", ""))
    time_ts = payload.get("time")
    if isinstance(time_ts, Mapping):
        msg.time.CopyFrom(_timestamp_message(time_ts))
    msg.totalUniqueUsers = int(payload.get("totalUniqueUsers", 0))
    for h in payload.get("uniqueUserHashes") or []:
        msg.uniqueUserHashes.append(int(h))
    instance_name = payload.get("instanceName")
    if instance_name:
        msg.instanceName = str(instance_name)
    app_version = payload.get("appVersion")
    if app_version:
        msg.appVersion = str(app_version)
    process_start = payload.get("processStartTime")
    if isinstance(process_start, Mapping):
        msg.processStartTime.CopyFrom(_timestamp_message(process_start))

    for stat in payload.get("stats") or []:
        if not isinstance(stat, Mapping):
            continue
        sm = msg.stats.add()
        sm.feature = str(stat.get("feature", ""))
        sm.uniqueContextIdentifierEnabledCount = int(
            stat.get("uniqueContextIdentifierEnabledCount", 0)
        )
        sm.uniqueContextIdentifierDisabledCount = int(
            stat.get("uniqueContextIdentifierDisabledCount", 0)
        )
        sm.uniqueUsersUsedCount = int(stat.get("uniqueUsersUsedCount", 0))
        for h in stat.get("uniqueUserHashes") or []:
            sm.uniqueUserHashes.append(int(h))
        for h in stat.get("uniqueViewedUserHashes") or []:
            sm.uniqueViewedUserHashes.append(int(h))
        variant_stats = stat.get("variantStats") or {}
        if isinstance(variant_stats, Mapping):
            for name, vs in variant_stats.items():
                if not isinstance(vs, Mapping):
                    continue
                entry = usage_pb2.VariantStats(  # type: ignore[attr-defined]
                    checkCount=int(vs.get("checkCount", 0)),
                    requestCount=int(vs.get("requestCount", 0)),
                    usedCount=int(vs.get("usedCount", 0)),
                    viewedCount=int(vs.get("viewedCount", 0)),
                )
                sm.variantStats[str(name)].CopyFrom(entry)
    return msg


def metric_stat_from_payload(payload: Mapping[str, Any]) -> Any:
    """Convert a batcher dict payload into a Metrics.MetricStat message."""
    from toggly.telemetry.pb import metrics_pb2

    msg = metrics_pb2.MetricStat()  # type: ignore[attr-defined]
    msg.appKey = str(payload.get("appKey", ""))
    msg.environment = str(payload.get("environment", ""))
    time_ts = payload.get("time")
    if isinstance(time_ts, Mapping):
        msg.time.CopyFrom(_timestamp_message(time_ts))
    instance_name = payload.get("instanceName")
    if instance_name:
        msg.instanceName = str(instance_name)

    def _fill_variant_values(target: Any, variant_values: Mapping[str, Any]) -> None:
        for name, value in variant_values.items():
            target.variantValues[str(name)] = float(value)

    for item in payload.get("stats") or []:
        if not isinstance(item, Mapping):
            continue
        sm = msg.stats.add()
        sm.metric = str(item.get("metric", ""))
        feature = item.get("feature")
        if feature:
            sm.feature = str(feature)
        vv = item.get("variantValues") or {}
        if isinstance(vv, Mapping):
            _fill_variant_values(sm, vv)

    for item in payload.get("counters") or []:
        if not isinstance(item, Mapping):
            continue
        cm = msg.counters.add()
        cm.metric = str(item.get("metric", ""))
        feature = item.get("feature")
        if feature:
            cm.feature = str(feature)
        vv = item.get("variantValues") or {}
        if isinstance(vv, Mapping):
            _fill_variant_values(cm, vv)

    for item in payload.get("observations") or []:
        if not isinstance(item, Mapping):
            continue
        om = msg.observations.add()
        obs_time = item.get("time")
        if isinstance(obs_time, Mapping):
            om.time.CopyFrom(_timestamp_message(obs_time))
        om.metric = str(item.get("metric", ""))
        feature = item.get("feature")
        if feature:
            om.feature = str(feature)
        vv = item.get("variantValues") or {}
        if isinstance(vv, Mapping):
            _fill_variant_values(om, vv)
    return msg


class _SharedChannel:
    def __init__(self, channel: Any) -> None:
        self._channel = channel
        self._closed = False

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._channel.close()


class _NativeUsageClient:
    def __init__(
        self,
        stub: Any,
        shared: _SharedChannel,
        default_metadata: Mapping[str, str],
        timeout: float,
    ) -> None:
        self._stub = stub
        self._shared = shared
        self._default_metadata = dict(default_metadata)
        self._timeout = timeout

    def send_stats(
        self,
        request: Mapping[str, Any],
        metadata: Mapping[str, str] | None = None,
    ) -> Any:
        meta = {**self._default_metadata, **(metadata or {})}
        msg = feature_stat_from_payload(request)
        return self._stub.SendStats(
            msg,
            metadata=tuple(meta.items()),
            timeout=self._timeout,
        )

    def close(self) -> None:
        self._shared.close()


class _NativeMetricsClient:
    def __init__(
        self,
        stub: Any,
        shared: _SharedChannel,
        default_metadata: Mapping[str, str],
        timeout: float,
    ) -> None:
        self._stub = stub
        self._shared = shared
        self._default_metadata = dict(default_metadata)
        self._timeout = timeout

    def send_metrics(
        self,
        request: Mapping[str, Any],
        metadata: Mapping[str, str] | None = None,
    ) -> Any:
        meta = {**self._default_metadata, **(metadata or {})}
        msg = metric_stat_from_payload(request)
        return self._stub.SendMetrics(
            msg,
            metadata=tuple(meta.items()),
            timeout=self._timeout,
        )

    def close(self) -> None:
        self._shared.close()


def create_grpc_clients(
    metrics_base_url: str,
    user_agent: str | None = None,
    *,
    timeout: float = 10.0,
) -> GrpcClients | None:
    """Dial usage + metrics stubs. Returns ``None`` when gRPC deps are missing."""
    if not is_grpc_available():
        return None

    import grpc

    from toggly.telemetry.pb import metrics_pb2_grpc, usage_pb2_grpc

    target = grpc_target(metrics_base_url)
    channel = grpc.secure_channel(target, grpc.ssl_channel_credentials())
    shared = _SharedChannel(channel)
    ua = resolve_user_agent(user_agent)
    default_meta = {GRPC_USER_AGENT_METADATA_KEY: ua}

    usage_stub = usage_pb2_grpc.UsageStub(channel)
    metrics_stub = metrics_pb2_grpc.MetricsStub(channel)

    return GrpcClients(
        usage=_NativeUsageClient(usage_stub, shared, default_meta, timeout),
        metrics=_NativeMetricsClient(metrics_stub, shared, default_meta, timeout),
    )


# Typing helper for injectable factories in tests
GrpcClientFactory = Callable[[str, Optional[str]], Optional[GrpcClients]]
